packages/dds/task-manager/src/test/taskManager.fuzz.spec.ts

```diff
@@ -0,0 +1,422 @@
+};
+
+function logCurrentState(state: FuzzTestState, loggingInfo: LoggingInfo): void {
+    for (const client of state.clients) {
+        const taskManager = client.taskManager;
+        assert(taskManager);
+        if (loggingInfo.taskManagerNames.includes(taskManager.id)) {
+            console.log(`TaskManager ${taskManager.id} (Connected: ${taskManager.connected}):`);
+            console.log(taskManager._getTaskQueues().get(loggingInfo.taskId));
+            console.log("\n");
+        }
+    }
+}
+
+function assertEqualQueues(queue1: Map<string, string[]>, queue2: Map<string, string[]>) {
+    assert(queue1.size === queue2.size, "The number of tasks queues are not the same");
+    for (const [key, val] of queue1) {
+        const testVal = queue2.get(key);
+        if (testVal === undefined) {
+            assert(val === undefined, "Task queues are not both undefined");
+            return;
+        }
+        assert(testVal.length === val.length, "Task queues are not the same size");
+        if (testVal.length > 0) {
+            testVal.forEach((task: string, index: number) => {
+                assert(task === val[index], `Task queues are not identical`);
+            });
+        }
+    }
+}
+
+function assertConsistent(clients: Client[]) {
+    const connectedClients = clients.filter((client) => client.containerRuntime.connected);
+    if (connectedClients.length < 2) {
+        return;
+    }
+    const first = connectedClients[0].taskManager;
+    for (const { taskManager: other } of connectedClients.slice(1)) {
+        assertEqualQueues(first._getTaskQueues(), other._getTaskQueues());
+    }
+}
+
+function runTaskManagerFuzz(
+    generator: Generator<Operation, FuzzTestState>,
+    initialState: FuzzTestState,
+    saveInfo?: SaveInfo,
+    loggingInfo?: LoggingInfo,
+): void {
+    // Small wrapper to avoid having to return the same state repeatedly; all operations in this suite mutate.
+    // Also a reasonable point to inject logging of incremental state.
+    const statefully =
+        <T>(statefulReducer: (state: FuzzTestState, operation: T) => void): Reducer<T, FuzzTestState> =>
+            (state, operation) => {
+                if (loggingInfo !== undefined && (operation as any).taskId === loggingInfo.taskId) {
+                    logCurrentState(state, loggingInfo);
+                    console.log("-".repeat(20));
+                    console.log("Next operation:", JSON.stringify(operation, undefined, 4));
+                }
+                statefulReducer(state, operation);
+                return state;
+            };
+
+    performFuzzActions(
+        generator,
+        {
+            volunteer: statefully(({ clients }, { taskManagerName, taskId }) => {
+                const taskManager = findTaskManager(clients, taskManagerName);
+                assert(taskManager);
+                taskManager.volunteerForTask(taskId).catch((e: Error) => {
```

Comment:
leaving uncaught, unawaited promises here seems undesirable. Do you think it would be valuable to put them in a "pending volunteer promises" list and somehow incorporate awaiting elements of that list into the test? I'm not familiar with the guarantees around the APIs, but it seems like either the synchronization step or upon test completion (at which point you could do something like disconnect clients) would suffice for those promises to resolve.

My main concern here is that the test actually fails with a well-reported issue if there's something going wrong in that promise. I believe Mocha adds an unhandled promise rejection error that makes things work if the promise happens to reject during the course of the test, but the error is worse if it happens after (or in the middle of a subsequent test or something).

Reply:
I think it depends on what these tests are supposed to represent -- if they're just acting as a "slice" of a collaboration, then it's expected some clients might be sitting with pending promises at any given time.  If instead they should represent a full collaboration session (i.e. all collaborators disconnect at the end) then they should settle.

Could potentially validate on the end of the test that each client with an unsettled promise (tracked like Abe suggests) are also .queued() and not .assigned(), though maybe just nice-to-have I think.

Nit: the catch block could be replaced with an assert.doesNotReject if you'd like though it's also fine as is.